sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,347 @@
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                return Equals(parameterSymbol, symbol);
+            }
+
+            public Diagnostic GetRuleViolation()
+            {
+                if (ShouldReportOn)
+                {
+                    var mostGeneralType = FindMostGeneralType();
+
+                    if (!Equals(mostGeneralType, parameterSymbol.Type) &&
+                        ShouldReportOnType(mostGeneralType.GetSymbolType()))
+                    {
+                        return Diagnostic.Create(rule,
+                            parameterSymbol.Locations.First(),
+                            mostGeneralType.ToDisplayString(), parameterSymbol.Type.ToDisplayString());
+                    }
+                }
+
+                return null;
+            }
+
+            private static bool ShouldReportOnType(ITypeSymbol typeSymbol)
+            {
+                return
+                       !typeSymbol.Is(KnownType.System_Object) &&
+                       !typeSymbol.Is(KnownType.System_ValueType) &&
+                       !typeSymbol.Name.StartsWith("_", StringComparison.InvariantCulture) &&
+                       !typeSymbol.Is(KnownType.System_Enum);
+            }
+
+            private ISymbol FindMostGeneralType()
+            {
+                var mostGeneralType = parameterSymbol.Type;
+
+                if (usedAs.Count == 0)
+                {
+                    return mostGeneralType;
+                }
+
+                mostGeneralType = FindMostGeneralClass(mostGeneralType);
+                mostGeneralType = FindMostGeneralInterface(mostGeneralType);
+                return mostGeneralType;
+            }
+
+            private ITypeSymbol FindMostGeneralClass(ITypeSymbol mostGeneralType)
+            {
+                ITypeSymbol currentSymbol = mostGeneralType.BaseType;
+
+                while (currentSymbol != null)
+                {
+                    if (DerivesOrImplementsAll(currentSymbol))
+                    {
+                        mostGeneralType = currentSymbol;
+                    }
+
+                    currentSymbol = currentSymbol?.BaseType;
+                }
+
+                return mostGeneralType;
+            }
+
+            private ITypeSymbol FindMostGeneralInterface(ITypeSymbol mostGeneralType)
+            {
+                foreach (var @interface in mostGeneralType.Interfaces)
+                {
+                    if (DerivesOrImplementsAll(@interface))
+                    {
+                        return FindMostGeneralInterface(@interface);
+                    }
+                }
+
+                return mostGeneralType;
+            }
+
+            private bool DerivesOrImplementsAll(ITypeSymbol type)
+            {
+                return type != null &&
+                    usedAs.All(u => type.DerivesOrImplements(u));
```

Comment:
With @valhristov we tend to prefer using the (no idea what is the proper name for it) syntax which doesn't specify the arrow when we can.  So `.All(u => type.DerivesOrImplements(u))` rewritten as `.All(type.DerivesOrImplements)`